Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,289 @@
+            // No data to send.
+            return nil
+        }
+
+        /// Once we return from this function, we want to account for the bytes we've handed off.
+        defer {
+            self.outboundBuffer.moveReaderIndex(to: self.outboundBuffer.writerIndex)
+        }
+
+        return self.outboundBuffer
+    }
+
+    /// Stores a buffer received from the network for delivery to OpenSSL.
+    ///
+    /// Whenever a buffer is received from the network, it is passed to the BIO via this function
+    /// call. Only one buffer may be passed to OpenSSL at any one time: once a buffer is passed, it
+    /// is expected to be cleared immediately. If it is not cleared, this is an application error and
+    /// must be resolved.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer of ciphertext bytes received from the network.
+    func receiveFromNetwork(buffer: ByteBuffer) {
+        precondition(self.inboundBuffer == nil, "Did not flush inbound bytes to OpenSSL")
+        self.inboundBuffer = buffer
+    }
+
+    /// OpenSSL has requested to read ciphertext bytes from the network.
+    ///
+    /// This function is invoked whenever OpenSSL is looking to read data.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer for NIO to copy bytes into.
+    /// - returns: The number of bytes we have copied.
+    fileprivate func sslRead(buffer: UnsafeMutableBufferPointer<CChar>) -> CInt {
+        guard var inboundBuffer = self.inboundBuffer else {
+            // We have no bytes to read. Mark this as "needs read retry".
+            CNIOOpenSSL_BIO_set_retry_read(self.bioPtr)
+            return -1
+        }
+
+        let bytesToCopy = min(buffer.count, inboundBuffer.readableBytes)
+        _ = inboundBuffer.readWithUnsafeReadableBytes { bytePointer in
+            assert(bytePointer.count >= bytesToCopy)
+            memcpy(buffer.baseAddress!, bytePointer.baseAddress!, bytesToCopy)
+            return bytesToCopy
+        }
+
+        // If we have read all the bytes from the inbound buffer, nil it out.
+        if inboundBuffer.readableBytes > 0 {
+            self.inboundBuffer = inboundBuffer
+        } else {
+            self.inboundBuffer = nil
+        }
+
+        return CInt(bytesToCopy)
+    }
+
+    /// OpenSSL has requested to write ciphertext bytes from the network.
+    ///
+    /// - parameters:
+    ///     - buffer: The buffer for NIO to copy bytes from.
+    /// - returns: The number of bytes we have copied.
+    fileprivate func sslWrite(buffer: UnsafeBufferPointer<CChar>) -> CInt {
+        if self.mustClearOutboundBuffer {
+            // We just flushed, and this is a new write. Let's clear the buffer now.
+            self.outboundBuffer.clear()
+            assert(!self.mustClearOutboundBuffer)
+        }
+
+        let byteCount = buffer.count
+        let writtenBytes = buffer.baseAddress!.withMemoryRebound(to: UInt8.self, capacity: byteCount) { (newPointer) -> Int in
```

Comment:
I don't think `withMemoryRebound(to:)` is correct here because you don't know the original type

Reply:
(in the new model)

Reply:
Yes I do, it's `CChar`. This pointer came in from OpenSSL as a `const char *`.